### ISSUE TITLE
IPv6 only

### ISSUE THREAD

**Issue by RJGhugo** [mastodon/mastodon#31395]
### Steps to reproduce the problem

The **IPv6 only** installation of Mastodon (Docker) seems to have several problems.

### Expected behaviour

run a single stack IPv6 server

### Actual behaviour

have to run a dual stack server or an IPv4 server

### Detailed description

It would be nice to test all new versions of Mastodon on an **IPv6 only** environment.

### Mastodon version

v4.x

**Comment by ThisIsMissEm** [comment]
This is probably a duplicate of #31199 and #25371

**Comment by RJGhugo** [comment]
I am not able to run Mastodon (4.2.11) on a **IPv6** only environment.

Is there a standard procedure to test new Mastodon versions with IPv6 only, before the are going in production?

**Comment by RJGhugo** [comment]
It seems to run now, somehow!

a) Bind (k3s)

**Workaround** as mentioned in https://github.com/mastodon/mastodon/issues/27043 and https://github.com/mastodon/mastodon/issues/25371.

`---`
`apiVersion: v1`
`kind: ConfigMap`
`metadata:`
`  name: bind`
`data:`
`  BIND: "[::]"`
`  REDIS_HOST: mastodon-redis`
`  REDIS_PORT: "6379"`
`---`
`apiVersion: v1`
`kind: ConfigMap`
`metadata:`
`  name: bind-stream`
`data:`
`  BIND: "::"`
`  REDIS_URL: "redis://mastodon-redis:6379/?family=6"`
`---`

b) Stream Log

**Why do I get a woring?**

`WARN Streaming API now listening on :::4000`

c) Web Log

`[7] Puma starting in cluster mode...`
`[7] * Puma version: 6.4.2 (ruby 3.2.3-p157) ("The Eagle of Durango")`
`[7] *  Min threads: 5`
`[7] *  Max threads: 5`
`[7] *  Environment: production`
`[7] *   Master PID: 7`
`[7] *      Workers: 2`
`[7] *     Restarts: (✔) hot (✖) phased`
`[7] * Preloading application`
`[7] * Listening on http://[::]:3000`
`[7] Use Ctrl-C to stop`
`[7]  Worker 0 (PID: 10) booted in 0.0s, phase: 0`
`[7] - Worker 1 (PID: 17) booted in 0.0s, phase: 0`

d) Sidekiq Log

`2024-08-18T14:06:27.351Z pid=7 tid=5en INFO: Booting Sidekiq 6.5.12 with Sidekiq::RedisConnection::RedisAdapter options {:driver=>:hiredis, :url=>"redis://mastodon-redis:6379/0", :namespace=>nil}`

**Comment by ThisIsMissEm** [comment]
Does `BIND: "[::]"` work for streaming?

**Comment by ThisIsMissEm** [comment]
In my testing, `BIND` of `[::]` does not work: 

```
{"level":"error","time":1723998095912,"pid":49446,"hostname":"the-beast-kitten.local","name":"streaming","err":{"type":"Error","message":"getaddrinfo ENOTFOUND [::]","stack":"Error: getaddrinfo ENOTFOUND [::]\n    at GetAddrInfoReqWrap.onlookup [as oncomplete] (node:dns:109:26)","errno":-3008,"code":"ENOTFOUND","syscall":"getaddrinfo","hostname":"[::]"},"msg":"getaddrinfo ENOTFOUND [::]"}
```

**Comment by ThisIsMissEm** [comment]
A potential fix for this discrepancy would be:

```diff
diff --git a/streaming/index.js b/streaming/index.js
index 8f66362178..86a4ef7efc 100644
--- a/streaming/index.js
+++ b/streaming/index.js
@@ -1563,7 +1563,13 @@ const attachServerWithConfig = (server, onSuccess) => {
       }
     });
   } else {
-    server.listen(+(process.env.PORT || 4000), process.env.BIND || '127.0.0.1', () => {
+    let bind = process.env.BIND ?? '127.0.0.1';
+
+    // Node.js doesn't handle `[::]` syntax, but does handle `::`
+    if (bind === '[::]') {
+      bind = '::';
+    }
+
+    server.listen(+(process.env.PORT || 4000), bind, () => {
       if (onSuccess) {
         onSuccess(`${server.address().address}:${server.address().port}`);
       }

```

**Comment by ThisIsMissEm** [comment]
Interestingly, `BIND='::'` works for node.js but fails for ruby/rails:

```
$ BIND="::" rails s
```

```
[49977] Puma starting in cluster mode...
[49977] * Puma version: 6.4.2 (ruby 3.3.4-p94) ("The Eagle of Durango")
[49977] *  Min threads: 5
[49977] *  Max threads: 5
[49977] *  Environment: development
[49977] *   Master PID: 49977
[49977] *      Workers: 2
[49977] *     Restarts: (✔) hot (✖) phased
[49977] * Preloading application
Exiting
/Users/emelia/.rbenv/versions/3.3.4/lib/ruby/3.3.0/uri/rfc3986_parser.rb:130:in `split': bad URI(is not URI?): "tcp://:::3000" (URI::InvalidURIError)
```

**Comment by RJGhugo** [comment]
I splitted the **ConfigMap**

`BIND: "[::]"` for web and sidekiq
`BIND: "::"` for stream
and
`REDIS_HOST: mastodon-redis` for web and sidekiq
`REDIS_PORT: "6379"`for web and sidekiq
`REDIS_URL: "redis://mastodon-redis:6379/?family=6"` for stream

see https://github.com/mastodon/mastodon/issues/27043 and https://github.com/mastodon/mastodon/issues/25371.

**Comment by ThisIsMissEm** [comment]
Have opened a fix in #31624
